pkg/kube/kube.go

```diff
@@ -603,17 +610,43 @@ func mergeObjects(live, obj runtime.Object) error {
 		wantPort := svc.Spec.HealthCheckNodePort
 		// If port is set (non-zero) and doesn't match the existing port (also non-zero), error out.
 		if wantPort != 0 && gotPort != 0 && wantPort != gotPort {
-			return fmt.Errorf("cannot update .spec.healthCheckNodePort to new value (want: %d, got: %d). requires resource recreation", wantPort, gotPort)
+			return true, errImmutableRessource(".spec.healthCheckNodePort")
 		}
 		svc.Spec.HealthCheckNodePort = gotPort
 	}
 
+	if liveClusterRoleBinding, ok := live.(*rbacv1.ClusterRoleBinding); ok {
+		clusterRoleBinding := obj.(*rbacv1.ClusterRoleBinding)
+		if liveClusterRoleBinding.RoleRef.APIGroup != clusterRoleBinding.RoleRef.APIGroup ||
+			liveClusterRoleBinding.RoleRef.Kind != clusterRoleBinding.RoleRef.Kind ||
+			liveClusterRoleBinding.RoleRef.Name != clusterRoleBinding.RoleRef.Name {
+			return true, errImmutableRessource("roleRef")
+		}
+	}
+
 	// Set metadata.resourceVersion for updates as required by
 	// Kubernetes API (http://go/k8s-concurrency).
 	if gotRV := live.(metav1.Object).GetResourceVersion(); gotRV != "" {
 		obj.(metav1.Object).SetResourceVersion(gotRV)
 	}
 
+	return false, nil
+}
+
+// maybeRecreate can be called to check if a resource can be updated or is immutable and needs recreation.
+// It evaluates if resource should be forcefully recreated, or if not, an error is thrown.
+func maybeRecreate(ctx context.Context, live, obj runtime.Object, m *kubePackage, r *apiResource) error {
```

Comment:
Perhaps the function docstring was written before the function impl, but the comments do not reflect the destructive nature (KubeDelete) of this function. Maybe also consider renaming this function to `mergeObjectsAndDeleteIfForce`.

Reply:
I updated the doc string and thought about the function name for a while. `mergeObjectsAndDeleteIfForce` doesn't really give it justice, cause it deletes AND recreates and ONLY IF it's an immutable attribute, which is rarely the case. The function name giving the full functionality would be something like `mergeObjectsAndDeleteAndRecreateIfImmutableAndForce` 🙈 I decided that the doc string and the code has to be warning enough if that's okay with you.